In [16]:
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
import seaborn as sns
%matplotlib inline

# For this homework is used laptops dataset

In [6]:
df=pd.read_csv('./data/2024_laptops.csv')

In [7]:
df

Laptop       Status   Brand  \
0     ASUS ExpertBook B1 B1502CBA-EJ0436X Intel Core...          New    Asus   
1     Alurin Go Start Intel Celeron N4020/8GB/256GB ...          New  Alurin   
2     ASUS ExpertBook B1 B1502CBA-EJ0424X Intel Core...          New    Asus   
3     MSI Katana GF66 12UC-082XES Intel Core i7-1270...          New     MSI   
4     HP 15S-FQ5085NS Intel Core i5-1235U/16GB/512GB...          New      HP   
...                                                 ...          ...     ...   
2155  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2156  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2157  Razer Blade 17 FHD 360Hz Intel Core i7-11800H/...  Refurbished   Razer   
2158  Razer Book 13 Intel Evo Core i7-1165G7/16GB/1T...  Refurbished   Razer   
2159  Razer Book FHD+ Intel Evo Core i7-1165G7/16GB/...  Refurbished   Razer   

           Model                CPU  RAM  Storage Storage type       GPU  \
0     ExpertBook      Intel Core i5    8      512          SSD       NaN   
1             Go      Intel Celeron    8      256          SSD       NaN   
2     ExpertBook      Intel Core i3    8      256          SSD       NaN   
3         Katana      Intel Core i7   16     1000          SSD  RTX 3050   
4            15S      Intel Core i5   16      512          SSD       NaN   
...          ...                ...  ...      ...          ...       ...   
2155       Blade      Intel Core i7   16     1000          SSD  RTX 3060   
2156       Blade      Intel Core i7   16     1000          SSD  RTX 3070   
2157       Blade      Intel Core i7   32     1000          SSD  RTX 3080   
2158        Book  Intel Evo Core i7   16     1000          SSD       NaN   
2159        Book  Intel Evo Core i7   16      256          SSD       NaN   

      Screen Touch  Final Price  
0       15.6    No      1009.00  
1       15.6    No       299.00  
2       15.6    No       789.00  
3       15.6    No      1199.00  
4       15.6    No       669.01  
...      ...   ...          ...  
2155    17.3    No      2699.99  
2156    17.3    No      2899.99  
2157    17.3    No      3399.99  
2158    13.4   Yes      1899.99  
2159    13.4   Yes      1699.99  

[2160 rows x 12 columns]

In [38]:
df.dtypes

Laptop           object
Status           object
Brand            object
Model            object
CPU              object
RAM               int64
Storage           int64
Storage type     object
GPU              object
Screen          float64
Touch            object
Final Price     float64
dtype: object

In [48]:
df.describe()

RAM      Storage       Screen  Final Price
count  2160.000000  2160.000000  2156.000000  2160.000000
mean     15.413889   596.294444    15.168112  1312.638509
std       9.867815   361.220506     1.203329   911.475417
min       4.000000     0.000000    10.100000   201.050000
25%       8.000000   256.000000    14.000000   661.082500
50%      16.000000   512.000000    15.600000  1031.945000
75%      16.000000  1000.000000    15.600000  1708.970000
max     128.000000  4000.000000    18.000000  7150.470000

In [50]:
df.describe().round(2)

RAM  Storage   Screen  Final Price
count  2160.00  2160.00  2156.00      2160.00
mean     15.41   596.29    15.17      1312.64
std       9.87   361.22     1.20       911.48
min       4.00     0.00    10.10       201.05
25%       8.00   256.00    14.00       661.08
50%      16.00   512.00    15.60      1031.95
75%      16.00  1000.00    15.60      1708.97
max     128.00  4000.00    18.00      7150.47

# Q1. Pandas version

What's the version of Pandas that you installed?

In [39]:
pd.__version__

'2.3.1'

In [58]:
# Correct answer
pd.__version__

'2.3.1'

# Q2. Records count

How many records are in the dataset?

- 12
- 1000
- 2160
- 12160

In [9]:
len(df)

2160

In [57]:
# Correct answer
df.shape[0]

2160

# Q3. Laptop brands

How many laptop brands are presented in the dataset?

- 12
- 27
- 28
- 2160

In [14]:
len(df.Brand.unique())

27

In [52]:
# Correct answer
df.Brand.nunique()

27

In [44]:
df.nunique()

Laptop          2160
Status             2
Brand             27
Model            121
CPU               28
RAM                9
Storage           12
Storage type       2
GPU               44
Screen            29
Touch              2
Final Price     1440
dtype: int64

# Q4. Missing values

How many columns in the dataset have missing values?

- 0
- 1
- 2
- 3

In [51]:
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

In [56]:
# Correct answer
df.isnull().sum()

Laptop             0
Status             0
Brand              0
Model              0
CPU                0
RAM                0
Storage            0
Storage type      42
GPU             1371
Screen             4
Touch              0
Final Price        0
dtype: int64

# Q5. Maximum final price

What's the maximum final price of Dell notebooks in the dataset?

- 869
- 3691
- 3849
- 3936

In [37]:
df[df.Brand == 'Dell']['Final Price'].max()

np.float64(3936.0)

In [54]:
# Correct answer
df[df.Brand == 'Dell']['Final Price'].max()

np.float64(3936.0)

In [47]:
df.groupby('Brand')['Final Price'].agg(['min', 'mean', 'max'])

min         mean      max
Brand                                          
Acer               264.14  1001.285766  3691.00
Alurin             239.00   484.701379   869.00
Apple              299.00  1578.227672  3849.00
Asus               239.25  1269.380699  5758.14
Deep Gaming       1334.00  1505.377500  1639.01
Dell               379.00  1153.839881  3936.00
Denver             329.95   329.950000   329.95
Dynabook Toshiba   397.29   999.197895  1805.01
Gigabyte           799.00  1698.488958  3799.00
HP                 210.14   952.628478  5368.77
Innjoo             251.40   347.888333   431.38
Jetwing            469.27   469.270000   469.27
LG                 609.04  1730.028438  2399.00
Lenovo             239.00  1087.723880  5018.14
MSI                444.19  1913.451006  7150.47
Medion             246.89   905.331875  3799.00
Microsoft          599.01  1595.543247  3747.91
Millenium         1640.98  1976.845000  2312.71
PcCom             1049.00  1606.762917  1949.90
Primux             208.99   339.618750   599.41
Prixton            246.95   292.280000   329.95
Razer             1232.74  3265.701622  4999.01
Realme             999.00   999.000000   999.00
Samsung            365.90  1595.511364  3699.01
Thomson            201.05   316.860000   436.56
Toshiba            799.00   799.000000   799.00
Vant               632.00  1047.455455  1217.01

# Q6. Median value of Screen

Find the median value of Screen column in the dataset.
Next, calculate the most frequent value of the same Screen column.
Use fillna method to fill the missing values in Screen column with the most frequent value from the previous step.
Now, calculate the median value of Screen once again.
Has it changed?

Hint: refer to existing mode and median functions to complete the task.

- Yes
- No

# Q7. Sum of weights

Select all the "Innjoo" laptops from the dataset.
Select only columns RAM, Storage, Screen.
Get the underlying NumPy array. Let's call it X.
Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
Compute the inverse of XTX.
Create an array y with values [1100, 1300, 800, 900, 1000, 1100].
Multiply the inverse of XTX with the transpose of X, and then multiply the result by y. Call the result w.
What's the sum of all the elements of the result?
Note: You just implemented linear regression. We'll talk about it in the next lesson.

- 0.43
- 45.29
- 45.58
- 91.30